src/main/kotlin/net/revanced/patcher/util/Io.kt

```diff
@@ -3,47 +3,90 @@ package net.revanced.patcher.util
+                // add it to our list
+                classes.add(classNode)
             }
+
+            // finally, close the entry
+            jis.closeEntry()
+        }
+
+        // at last reset the buffered input stream
+        bufferedInputStream.reset()
     }
 
-    fun writeClassesToJar(input: InputStream, output: OutputStream, classes: List<ClassNode>) {
-        val jis = JarInputStream(input)
-        val jos = JarOutputStream(output)
-
-        // TODO: Add support for adding new/custom classes
-        while (true) {
-            val next = jis.nextJarEntry ?: break
-            val e = JarEntry(next) // clone it, to not modify the input (if possible)
-            jos.putNextEntry(e)
-
-            val clazz = classes.singleOrNull {
-                clazz -> clazz.name+".class" == e.name // clazz.name is the class name only while e.name is the full filename with extension
-            };
-            if (clazz != null) {
-                val cw = ClassWriter(ClassWriter.COMPUTE_MAXS or ClassWriter.COMPUTE_FRAMES)
-                clazz.accept(cw)
-                jos.write(cw.toByteArray())
-            } else {
-                jos.write(jis.readBytes())
-            }
+    fun saveAsJar() {
+        val jis = ZipInputStream(bufferedInputStream)
+        val jos = ZipOutputStream(output)
+
+        // first write all non .class zip entries from the original input stream to the output stream
+        // we read it first to close the input stream as fast as possible
+        var zipEntry: ZipEntry?
+        while (jis.nextEntry.also { zipEntry = it } != null) {
+            // skip all class files because we added them in the loop above
+            // TODO: check for zipEntry.isDirectory
+            if (zipEntry!!.name.endsWith(".class")) continue
+
+            // create a new zipEntry and write the contents of the zipEntry to the output stream
+            jos.putNextEntry(ZipEntry(zipEntry!!.name))
+            jos.write(jis.readBytes())
+
+            // close the newly created zipEntry
+            jos.closeEntry()
+        }
+
+        // finally, close the input stream
+        jis.close()
+
+        // now write all the patched classes to the output stream
+        for (patchedClass in classes) {
+            // create a new entry of the patched class
+            jos.putNextEntry(JarEntry(patchedClass.name + ".class"))
+
+            // parse the patched class to a byte array and write it to the output stream
+            val cw = ClassWriter(ClassWriter.COMPUTE_MAXS or ClassWriter.COMPUTE_FRAMES)
+            patchedClass.accept(cw)
+            jos.write(cw.toByteArray())
+
+            // close the newly created jar entry
             jos.closeEntry()
         }
+
+        // finally, close all streams
+        // TODO: order & necessity to close all of the streams check
+        jos.close()
+        bufferedInputStream.close()
+        input.close()
```

Comment:
This one too. Place it under the `bufferedInputStream.close()` above the for-loop again.